// tslint-sonarts/lib/rules/noEmptyNestedBlocksRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
 * SonarTS
 * Copyright (C) 2017-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
const tslint = require("tslint");
const ts = require("typescript");
const navigation_1 = require("../utils/navigation");
class Rule extends tslint.Rules.AbstractRule {
    apply(sourceFile) {
        return this.applyWithWalker(new Walker(sourceFile, this.getOptions()));
    }
}
Rule.metadata = {
    ruleName: "no-empty-nested-blocks",
    description: "Nested blocks of code should not be left empty",
    rationale: tslint.Utils.dedent `
      Most of the time a block of code is empty when a piece of code is really missing. So such empty block must be
      either filled or removed.`,
    optionsDescription: "",
    options: null,
    rspecKey: "RSPEC-108",
    type: "functionality",
    typescriptOnly: false,
};
Rule.MESSAGE = "Either remove or fill this block of code.";
exports.Rule = Rule;
class Walker extends tslint.RuleWalker {
    visitBlock(node) {
        if (node.statements.length === 0 &&
            !this.hasComments(node) &&
            !this.isLikeFunction(node.parent) &&
            !this.isCatchClause(node.parent) &&
            !this.hasLeadingComment(node.parent)) {
            this.addFailureAtNode(node, Rule.MESSAGE);
        }
        super.visitBlock(node);
    }
    isLikeFunction(node) {
        return (node != null &&
            navigation_1.is(node, ts.SyntaxKind.FunctionDeclaration, ts.SyntaxKind.FunctionExpression, ts.SyntaxKind.Constructor, ts.SyntaxKind.MethodDeclaration, ts.SyntaxKind.GetAccessor, ts.SyntaxKind.SetAccessor, ts.SyntaxKind.ArrowFunction));
    }
    isCatchClause(node) {
        return node != null && node.kind === ts.SyntaxKind.CatchClause;
    }
    hasComments(node) {
        if (node.getChildCount() > 0) {
            const openBrace = node.getChildAt(0);
            const closeBrace = node.getChildAt(node.getChildCount() - 1);
            return navigation_1.getCommentsAfter(openBrace).length > 0 || navigation_1.getCommentsBefore(closeBrace).length > 0;
        }
        else {
            return false;
        }
    }
    hasLeadingComment(node) {
        return node != null && navigation_1.getCommentsBefore(node).length > 0;
    }
}
